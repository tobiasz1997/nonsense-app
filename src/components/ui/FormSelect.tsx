import cx from 'classnames';
import {
	forwardRef,
	ForwardRefRenderFunction,
	SelectHTMLAttributes
} from 'react';
import FormErrorMessage from '@components/ui/FormErrorMessage';
import FormLabel from '@components/ui/FormLabel';
import { SelectOption } from '@interfaces/selectOption';

type Props = {
	label?: string;
	error?: string;
	options: SelectOption[];
};

const FormSelect: ForwardRefRenderFunction<
	HTMLSelectElement,
	Props & SelectHTMLAttributes<HTMLSelectElement>
> = (props, ref) => {
	return (
		<div>
			<label
				className={cx(
					'flex cursor-pointer items-center overflow-hidden rounded border focus-within:ring-4',
					props.error
						? 'border-red text-red focus-within:ring-red/10'
						: 'border-green-dark text-green-dark focus-within:ring-yellow'
				)}
			>
				<span className="relative w-full">
					{props.label && props.label && (
						<span className="absolute top-1 left-3">
							<FormLabel required={props.required}>{props.label}</FormLabel>
						</span>
					)}

					<select
						{...props}
						ref={ref}
						className={cx(
							'h-12 w-full border-none pr-10 focus:outline-none focus:ring-0 disabled:opacity-50',
							props.label ? 'px-3 pb-2 pt-4' : 'p-2'
						)}
					>
						{props.options.map((option, index) => (
							<option key={index} value={option.value}>
								{option.label}
							</option>
						))}
					</select>
				</span>
			</label>
			{props.error && <FormErrorMessage>{props.error}</FormErrorMessage>}
		</div>
	);
};

export default forwardRef(FormSelect);
