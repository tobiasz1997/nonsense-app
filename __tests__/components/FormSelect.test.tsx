import { render } from '@testing-library/react';
import FormSelect from '@components/ui/FormSelect';
import { SelectOption } from '@interfaces/selectOption';

describe('Form Select', () => {
	const setup = () => {
		const options: SelectOption[] = [
			{
				label: 'First',
				value: 1
			},
			{
				label: 'Second',
				value: 2
			}
		];

		return {
			options
		};
	};

	it('should render without crashing', () => {
		const { options } = setup();
		const { container } = render(<FormSelect options={options} />);
		expect(container).toMatchSnapshot();
	});
});
