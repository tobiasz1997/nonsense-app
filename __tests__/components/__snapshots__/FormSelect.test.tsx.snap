// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Select should render without crashing 1`] = `
<div>
  <div>
    <label
      class="flex cursor-pointer items-center overflow-hidden rounded border focus-within:ring-4 border-green-dark text-green-dark focus-within:ring-yellow"
    >
      <span
        class="relative w-full"
      >
        <select
          class="h-12 w-full border-none pr-10 focus:outline-none focus:ring-0 disabled:opacity-50 p-2"
          options="[object Object],[object Object]"
        >
          <option
            value="1"
          >
            First
          </option>
          <option
            value="2"
          >
            Second
          </option>
        </select>
      </span>
    </label>
  </div>
</div>
`;
